src/main/java/net/minecraftforge/common/data/ForgeRecipeProvider.java

```diff
@@ -99,6 +96,100 @@ protected void buildCraftingRecipes(Consumer<FinishedRecipe> consumer)
+        addMixingRecipe(consumer, Potions.LONG_SWIFTNESS, Items.FERMENTED_SPIDER_EYE, Potions.LONG_SLOWNESS);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.SUGAR, Potions.SWIFTNESS);
+        addMixingRecipe(consumer, Potions.SWIFTNESS, Tags.Items.DUSTS_REDSTONE, Potions.LONG_SWIFTNESS);
+        addMixingRecipe(consumer, Potions.SWIFTNESS, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_SWIFTNESS);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.PUFFERFISH, Potions.WATER_BREATHING);
+        addMixingRecipe(consumer, Potions.WATER_BREATHING, Tags.Items.DUSTS_REDSTONE, Potions.LONG_WATER_BREATHING);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.GLISTERING_MELON_SLICE, Potions.HEALING);
+        addMixingRecipe(consumer, Potions.HEALING, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_HEALING);
+        addMixingRecipe(consumer, Potions.HEALING, Items.FERMENTED_SPIDER_EYE, Potions.HARMING);
+        addMixingRecipe(consumer, Potions.STRONG_HEALING, Items.FERMENTED_SPIDER_EYE, Potions.STRONG_HARMING);
+        addMixingRecipe(consumer, Potions.HARMING, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_HARMING);
+        addMixingRecipe(consumer, Potions.POISON, Items.FERMENTED_SPIDER_EYE, Potions.HARMING);
+        addMixingRecipe(consumer, Potions.LONG_POISON, Items.FERMENTED_SPIDER_EYE, Potions.HARMING);
+        addMixingRecipe(consumer, Potions.STRONG_POISON, Items.FERMENTED_SPIDER_EYE, Potions.STRONG_HARMING);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.SPIDER_EYE, Potions.POISON);
+        addMixingRecipe(consumer, Potions.POISON, Tags.Items.DUSTS_REDSTONE, Potions.LONG_POISON);
+        addMixingRecipe(consumer, Potions.POISON, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_POISON);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.GHAST_TEAR, Potions.REGENERATION);
+        addMixingRecipe(consumer, Potions.REGENERATION, Tags.Items.DUSTS_REDSTONE, Potions.LONG_REGENERATION);
+        addMixingRecipe(consumer, Potions.REGENERATION, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_REGENERATION);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.BLAZE_POWDER, Potions.STRENGTH);
+        addMixingRecipe(consumer, Potions.STRENGTH, Tags.Items.DUSTS_REDSTONE, Potions.LONG_STRENGTH);
+        addMixingRecipe(consumer, Potions.STRENGTH, Tags.Items.DUSTS_GLOWSTONE, Potions.STRONG_STRENGTH);
+        addMixingRecipe(consumer, Potions.WATER, Items.FERMENTED_SPIDER_EYE, Potions.WEAKNESS);
+        addMixingRecipe(consumer, Potions.WEAKNESS, Tags.Items.DUSTS_REDSTONE, Potions.LONG_WEAKNESS);
+        addMixingRecipe(consumer, Potions.AWKWARD, Items.PHANTOM_MEMBRANE, Potions.SLOW_FALLING);
+        addMixingRecipe(consumer, Potions.SLOW_FALLING, Tags.Items.DUSTS_REDSTONE, Potions.LONG_SLOW_FALLING);
+
+    }
+
+    private void addMixingRecipe(final Consumer<FinishedRecipe> consumer, final Potion input, final Item ingredient, final Potion output)
```

Comment:
Could make this protected for extension use.

Reply:
why ForgeRecipeProvider is final

Reply:
In that case then, this should probably be moved to some helper interface then for users to extend similar to `IConditionBuilder`. Makes relatively no sense to make these methods unobtainable to an average modder since they will be creating these jsons too.